src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,218 @@
+     */
+    private int singleSpaceAsteriskColNumber;
+
+    /** Specifies the lines of the file being processed. */
+    private String[] fileLines;
+
+    @Override
+    public int[] getDefaultJavadocTokens() {
+        return new int[] {
+            JavadocTokenTypes.LEADING_ASTERISK,
+        };
+    }
+
+    @Override
+    public int[] getRequiredJavadocTokens() {
+        return getAcceptableJavadocTokens();
+    }
+
+    @Override
+    public void beginJavadocTree(DetailNode rootAst) {
+        // this method processes and sets information of starting javadoc tag.
+        fileLines = getLines();
+        final String startLine = fileLines[rootAst.getLineNumber() - 1];
+        commentBeginLineNumber = rootAst.getLineNumber();
+        commentBeginColNumber = CommonUtil.lengthExpandedTabs(
+            startLine, rootAst.getColumnNumber() - 1, getTabWidth());
+    }
+
+    @Override
+    public void visitJavadocToken(DetailNode ast) {
+        // this method checks the alignment of leading asterisks.
+        if (ast.getLineNumber() != commentBeginLineNumber) {
+            final int asteriskColNumber = getAsteriskColumnNumberWithTabsExpanded(ast.getText());
+
+            checkAsteriskAlignment(
+                    ast.getLineNumber(),
+                    commentBeginColNumber,
+                    asteriskColNumber);
+        }
+    }
+
+    @Override
+    public void finishJavadocTree(DetailNode rootAst) {
+        // this method checks the alignment of closing javadoc tag.
+        final DetailAST javadocEndToken = getBlockCommentAst().getLastChild();
+        final String lastLine = fileLines[javadocEndToken.getLineNo() - 1];
+        final int asteriskColNumber = getAsteriskColumnNumberWithTabsExpanded(lastLine);
+
+        checkAsteriskAlignment(
+                javadocEndToken.getLineNo(),
+                commentBeginColNumber,
+                asteriskColNumber);
+    }
+
+    /**
+     * Processes and returns the column number of leading asterisk from javadoc comment text
+     * with tabs expanded.
+     * Also sets 'singleSpaceAsteriskColNumber' if the leading asterisk is found.
+     *
+     * @param line single line from block comment text
+     * @return column number of leading asterisk with tabs expanded
+     */
+    private int getAsteriskColumnNumberWithTabsExpanded(String line) {
+        int columnNumber = getAsteriskColumnNumber(line);
+
+        if (columnNumber != -1) {
+            singleSpaceAsteriskColNumber = columnNumber;
+            columnNumber = CommonUtil.lengthExpandedTabs(line, columnNumber, getTabWidth());
+        }
+
+        return columnNumber;
+    }
+
+    /**
+     * Processes and returns the column number of leading asterisk without tabs expanded.
+     * Returns -1 if there is no leading asterisk found.
+     *
+     * @param line javadoc comment line
+     * @return asterisk's column number
+     */
+    private static int getAsteriskColumnNumber(String line) {
```

Comment:
Hey. I've seen your reviews but my exams are going on so this new changes might take some time. But I'll try to implement them as soon as possible. Sorry for the delay

Reply:
We are all busy people, take your time

Reply:
> Please do:
> 
> * return an Optional here to make it more clear that we will not always have a leading asterisk
> * leave a concise comment inside of this method (not in the javadoc) why we may not always have a leading asterisk
> * Completely avoid dealing with `-1`'s throughout this check by simply taking no action if this method's return value is empty

Done. Here is the new implementation:

https://github.com/checkstyle/checkstyle/blob/d022231628421f2f56ac5ef1e0afd3779a9e67e3/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L184-L187


I have used Optional here and every other places which was dependent on this method. I used `OptionalInt` instead of `Optional<Integer>` because some CIs were giving this error:

```
[ERROR] Low: Method com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocLeadingAsteriskAlignCheck.getAsteriskColumnNumber(String) uses a java.util.Optional when use of OptionalInt, OptionalLong, OptionalDouble would be more clear [com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocLeadingAsteriskAlignCheck] At JavadocLeadingAsteriskAlignCheck.java:[line 196] OI_OPTIONAL_ISSUES_PRIMITIVE_VARIANT_PREFERRED
[ERROR] Low: Method com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocLeadingAsteriskAlignCheck.getAsteriskColumnNumberWithTabsExpanded(String) uses a java.util.Optional when use of OptionalInt, OptionalLong, OptionalDouble would be more clear [com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocLeadingAsteriskAlignCheck] At JavadocLeadingAsteriskAlignCheck.java:[line 173] OI_OPTIONAL_ISSUES_PRIMITIVE_VARIANT_PREFERRED
```

Reply:
I do not see any CI failures for using `Optional<Integer>` here: https://github.com/checkstyle/checkstyle/pull/14897

Please use changes from https://github.com/checkstyle/checkstyle/pull/14897/commits/c251b83b89364ee243a3aa946072778752d01738 and update this PR so that we can get it across the finish line.

Reply:
> Please use changes from https://github.com/checkstyle/checkstyle/commit/c251b83b89364ee243a3aa946072778752d01738 and update this PR so that we can get it across the finish line.

Done

> Please always link to failing CI run in the future.

Yess I forgot to do it this time.

Now everything seems to be working fine, Idk why it was showing errors earlier maybe I made some kind of mistake back then